sigstore/_cli.py

```diff
@@ -775,11 +775,37 @@ def _collect_verification_state(
                 missing.append(str(cert))
             input_map[file] = {"cert": cert, "sig": sig}
         else:
+            # NOTE(ww): If the user hasn't specified a bundle via `--bundle` and
+            # `{input}.sigstore.json` doesn't exist, then we try `{input}.sigstore`
+            # for backwards compatibility.
+            legacy_default_bundle = file.parent / f"{file.name}.sigstore"
+
             # If a user hasn't explicitly supplied `--signature`, `--certificate` or
             # `--rekor-bundle`, we expect a bundle either supplied via `--bundle` or with the
             # default `{input}.sigstore` name.
             if not bundle.is_file():
-                missing.append(str(bundle))
+                if legacy_default_bundle.is_file():
+                    logger.warning(
+                        f"{file}: {legacy_default_bundle} should be named {bundle}. "
+                        "Support for discovering 'bare' .sigstore inputs will be deprecated in "
+                        "a future release."
+                    )
+                    bundle = legacy_default_bundle
+                else:
+                    # NOTE: Logically `{input}.sigstore` is what's missing here, but
+                    # we only fail this check if the the `--bundle` or `{input}.sigstore.json`
+                    # forms are *also* missing. So we use the former in the missing set,
+                    # since it'll result in a more useful error message.
+                    missing.append(str(bundle))
+            else:
+                # Don't allow the user to implicitly verify `{input}.sigstore.json` if
+                # `{input}.sigstore` is also present, since this implies user confusion.
+                if bundle != legacy_default_bundle and legacy_default_bundle.is_file():
```

Comment:
NB: We need this inequality check because a user can pass in `--bundle` with the same value as `legacy_default_bundle`, which is what the conformance suite does.